libcrux-ml-dsa/src/matrix.rs

```diff
@@ -21,9 +22,25 @@ pub(crate) fn expand_to_A<
     let mut A = [[PolynomialRingElement::<SIMDUnit>::ZERO(); COLUMNS_IN_A]; ROWS_IN_A];
 
     // Mutable iterators won't go through hax, so we need these range loops.
+
+    // | Key size | ROWS_IN_A | COLUMNS_IN_A |
+    // | -------- | --------- | ------------ |
+    // | 44       | 4         | 4            |
+    // | 65       | 6         | 5            |
+    // | 87       | 8         | 7            |
+    //
+    // We always do 4 in parallel first and then one at a time.
     #[allow(clippy::needless_range_loop)]
     for i in 0..ROWS_IN_A {
-        for j in 0..COLUMNS_IN_A {
+        let samples = sample_four_ring_element_uniform::<SIMDUnit, Shake128X4>(seed, i);
+        A[i][0] = samples.0;
+        A[i][1] = samples.1;
+        A[i][2] = samples.2;
+        A[i][3] = samples.3;
+
+        // TODO: We could do more in parallel if we think that may speed
```

Comment:
The 4x4 and 8x7 cases could be fully parallelized in groups of 4, 5x6 is the odd one out, with at least to individual SHAKE calls. Perhaps in this case we could also do everything in groups of 4 and swallow a useless computation of two hashes, in case it's still faster than two individual calls? I'll make a note to check it out later.

Reply:
Yes, we can definitely do a little more in parallel here! Let's get a first round in and then do more.